Add tracks fundamentals (#90)
napari/tutorials: fundamentals/tracks.md

These properties can be used to set the colors of the tracks.

For example, when displaying tracks of different classes/types,
one could automatically set color the individual tracks by their respective class/type.

## a simple example

You can create a new viewer and add a set of tracks in one go using the `napari.view_tracks` method,
or if you already have an existing viewer,
you can add tracks to it using `viewer.add_tracks`.
The api of both methods is the same.

In this example of we will overlay some tracks on an image from the Hubble space telescope:

```python
import napari
from skimage import data

hubble_image = data.hubble_deep_field()

tracks_data = [
    [1, 0, 236, 0],
    [1, 1, 236, 100],
    [1, 2, 236, 200],
    [1, 3, 236, 500],
    [1, 4, 236, 1000],
    [2, 0, 436, 0],
    [2, 1, 436, 100],
    [2, 2, 436, 200],
    [2, 3, 436, 500],
    [2, 4, 436, 1000],
    [3, 0, 636, 0],
    [3, 1, 636, 100],
    [3, 2, 636, 200],
    [3, 3, 636, 500],
    [3, 4, 636, 1000]
]

with napari.gui_qt():
    viewer = napari.view_image(test_im, name='image')
    viewer.add_tracks(tracks_data, name='tracks')

```

> ![image]({{ '/assets/tutorials/tracks_simple_demo.gif' | relative_url }})

Reviewer: I love this demo!

tail_width : float
    Width of the track tails in pixels.
tail_length : float
    Length of the track tails in units of time.
colormap : str
    Default colormap to use to set vertex colors. Specialized colormaps,
    relating to specified properties can be passed to the layer via
    colormaps_dict.
colormaps_dict : dict {str: napari.utils.Colormap}
    Optional dictionary mapping each property to a colormap for that
    property. This allows each property to be assigned a specific colormap,
    rather than having a global colormap for everything.
name : str
    Name of the layer.
metadata : dict
    Layer metadata.
scale : tuple of float
    Scale factors for the layer.
translate : tuple of float
    Translation values for the layer.
opacity : float
    Opacity of the layer visual, between 0.0 and 1.0.
blending : str
    One of a list of preset blending modes that determines how RGB and
    alpha values of the layer visual get mixed. Allowed values are
    {'opaque', 'translucent', and 'additive'}.
visible : bool
    Whether the layer visual is currently being displayed.

Returns
-------
layer : napari.layers.Tracks
    The newly-created tracks layer.
"""
```

## tracks data

> The input data to the tracks layer must be an NxD numpy array

Reviewer: Should we change this now that the `Tracks` layer can accept a list, dataframe or array for `data`?

Author: Great point. Fixed!

containing the coordinates of N verticies with a track ID and coordinate in D dimensions. 
The ordering of these dimensions is the same as the ordering of the dimensions for image layers.
This array is always accessible through the `layer.data` property
and will grow or shrink as new tracks are either added or deleted.

Consider a set of 4D tracks data (time + 3 spatial dimensions) defined below. The `Tracks` layer assumes the first column is the `track_id`, the second column is the time axis, and columns 3-5 are Z, Y, and X, respectively. Each row is one vertex in a track. All vertices with the same `track_id` are joined into a single track. In this case, we have defined 2 tracks: track 0, which goes from [10, 10, 10] to [20, 10, 10] and track 1, which goes from [10, 8, 5] to [7, 8, 10] (coordinates written as [x, y z]\).

| track_id | t | z  | y  | x  |
|----------|---|----|----|----|
| 0        | 0 | 10 | 10 | 10 |
| 0        | 1 | 10 | 10 | 20 |
| 1        | 0 | 5  | 8  | 10 |
| 1        | 1 | 10 | 8  | 7  |

> We can pass the example data above to the tracks layer as follows:

Reviewer: ```suggestion
The data in the array must be sorted by increasing `track_id` then time, as shown above.  We can pass the example data above to the tracks layer as follows:
```

```python

tracks_data = [
    [0, 0, 10, 10, 10],
    [0, 1, 10, 10, 20],
    [1, 0, 5, 8, 10],
    [1, 1, 10, 8, 7]
]

with napari.gui_qt():
    viewer = napari.view_tracks(tracks_data)

```

## tracks graph

> We can use the tracks `graph` argument to define the relationships between tracks (e.g., tracks merging or tracks splitting. The graph is defined as a dictionary where the keys are the `track_id` and the values are the `track_id` of the parents of the the key.

Reviewer: ```suggestion
We can use the tracks `graph` argument to define the relationships between tracks (e.g., tracks merging or tracks splitting). The graph is defined as a dictionary where the keys are the `track_id` and the values are the `track_id` of the parents of the the key.
```

For example, if we have a track 0, which splits into tracks 1 and 2 (i.e., track 0 is the parent of tracks 1 and 2), we would define the graph as:

```python
graph = {
    1: [0],
    2: [0]
}
```
If later tracks 1 and 2 merge into track 3 (i.e,. tracks 1 and 2 are the parent of track 3), the dictionary would become

```python
graph = {
    1: [0],
    2: [0],
    3: [1, 2]  
}
```
For a full example of 3d+t tracks data with a parent graph, please see our [`tracks_3d_with_graph.py` example](https://github.com/napari/napari/blob/master/examples/tracks_3d_with_graph.py).

## using the tracks properties dictionary

> The `Tracks` layer can contain properties that annotate each point.

Reviewer: ```suggestion
The `Tracks` layer can contain properties that annotate the vertices of each track.
```
